source/portable/posix/core_pkcs11_pal.c

```diff
@@ -0,0 +1,418 @@
+                        "The PAL was unable to open a file with name %s in write mode.", pcFileName ) );
+            xHandle = ( CK_OBJECT_HANDLE ) eInvalidHandle;
+        }
+        else
+        {
+            ulBytesWritten = fwrite( pucData, sizeof( uint8_t ), ulDataSize, pxFile );
+
+            if( ulBytesWritten != ulDataSize )
+            {
+                LogError( ( "PKCS #11 PAL was unable to save object to file. "
+                            "Expected to write %lu bytes, but wrote %lu bytes.", ulDataSize, ulBytesWritten ) );
+                xHandle = ( CK_OBJECT_HANDLE ) eInvalidHandle;
+            }
+            else
+            {
+                LogDebug( ( "Successfully wrote %lu to %s", ulBytesWritten, pcFileName ) );
+            }
+        }
+
+        if( NULL != pxFile )
+        {
+            ( void ) fclose( pxFile );
+        }
+    }
+    else
+    {
+        LogError( ( "Could not save object. Unable to find the correct file." ) );
+    }
+
+    return xHandle;
+}
+
+/*-----------------------------------------------------------*/
+
+
+CK_OBJECT_HANDLE PKCS11_PAL_FindObject( CK_BYTE_PTR pxLabel,
+                                        CK_ULONG usLength )
+{
+    const char * pcFileName = NULL;
+    CK_OBJECT_HANDLE xHandle = ( CK_OBJECT_HANDLE ) eInvalidHandle;
+
+    ( void ) usLength;
+
+    if( pxLabel != NULL )
+    {
+        prvLabelToFilenameHandle( ( const char * ) pxLabel,
+                                  &pcFileName,
+                                  &xHandle );
+
+        if( CKR_OK != prvFileExists( pcFileName ) )
+        {
+            xHandle = ( CK_OBJECT_HANDLE ) eInvalidHandle;
+        }
+    }
+    else
+    {
+        LogError( ( "Could not find object. Received a NULL label." ) );
+    }
+
+    return xHandle;
+}
+/*-----------------------------------------------------------*/
+
+CK_RV PKCS11_PAL_GetObjectValue( CK_OBJECT_HANDLE xHandle,
+                                 CK_BYTE_PTR * ppucData,
+                                 CK_ULONG_PTR pulDataSize,
+                                 CK_BBOOL * pIsPrivate )
+{
+    CK_RV xReturn = CKR_OK;
+    const char * pcFileName = NULL;
+
+
+    if( ( ppucData == NULL ) || ( pulDataSize == NULL ) || ( pIsPrivate == NULL ) )
+    {
+        xReturn = CKR_ARGUMENTS_BAD;
+        LogError( ( "Could not get object value. Received a NULL argument." ) );
+    }
+    else
+    {
+        xReturn = prvHandleToFilename( xHandle, &pcFileName, pIsPrivate );
+    }
+
+    if( xReturn == CKR_OK )
+    {
+        xReturn = prvReadData( pcFileName, ppucData, pulDataSize );
+    }
+
+    return xReturn;
+}
+
+/*-----------------------------------------------------------*/
+
+void PKCS11_PAL_GetObjectValueCleanup( CK_BYTE_PTR pucData,
+                                       CK_ULONG ulDataSize )
+{
+    /* Unused parameters. */
+    ( void ) ulDataSize;
+
+    if( NULL != pucData )
+    {
+        PKCS11_FREE( pucData );
```

Comment:
The object value is allocated using `malloc` but uses `PKCS11_FREE` here. Should this be consistent?

Reply:
Oops, thanks for catching that, just pushed a fix